### DISCUSSION TITLE
Displacement output

### DISCUSSION THREAD

**Discussion by CsatiZoltan** [neperfepx/FEPX#19] [Q&A] [answered]
Hi,

In section *2.1.6 Printing Results* of the documentation, the nodal coordinates and velocities are available but not the displacements. I guess that this is because FEPX solves a time-dependent (elasto-*visco*-plastic) problem. However, I would need to have access to the displacements as well, which can easily be determined from the nodal coordinates. Is there a reason why it is not implemented in FEPX? If it will soon be available in a future release, I may wait to implement it by myself...

Thank you,
Zoltan

**Comment by mkasemer** [comment, marked as answer]
@CsatiZoltan - you are right that it is relatively easy to calculate displacements, either in-code or during post-processing. We are working on a new release for sometime early next week (the plan was today, but a bug was found). I can attempt to add displacement output to the next release, but otherwise I will let you know, and you could submit a pull request with changes. I assume you would prefer total (cumulative) displacements, rather than incremental?

In the meantime, if it cannot wait until Monday or so, Neper should have the ability to perform such mathematical manipulations during post-processing. @rquey could comment here.

**Reply by CsatiZoltan** [reply]
The reason I would prefer it in-code and make it available as a requested output is that in that case no post-processing would be needed<sup> 1 </sup>. I think that would eventually be useful for other users as well (many crystal plasticity frameworks neglect viscosity effects). I would be happy to contribute in the form of a PR, but unfortunately I do not know Fortran. Tell me how I could help!
Yes, I would prefer total displacements.

If displacements do not get incorporated into the upcoming release, I would be glad if a post-processing solution with Neper was provided. I could implement it in Python by tinkering with the output files, but that would need to take into account the distributed nature of the solution (i.e. fetch results from the different cores). Surely doable, but a built-in solution was more elegant.

<sup>1</sup> This would make it independent of Neper. I recently installed FEPX on a cluster. It was easy as FEPX has few dependencies. On the other hand, Neper was more difficult due to the several dependencies, the satisfaction of which needs admin rights. Finally, I ended up with creating a `conda` environment and installing Neper in there.

**Reply by mkasemer** [reply]
@CsatiZoltan: I edited my comment above to give more context / an expected timeline. I hope early next week is acceptable.

@rquey can comment both on Neper's post-processing capabilities (and you can see some examples at https://neper.info), as well as some upcoming updates which may make it easier to install on a cluster without admin access.

**Reply by CsatiZoltan** [reply]
@mkasemer Thanks. Yes, I can wait for the next release, no problem.

> I can attempt to add displacement output to the next release

Is there a roadmap available about upcoming features of FEPX, or that is private? I understand though if the license does not permit publishing it publicly.

**Reply by mkasemer** [reply]
@CsatiZoltan: I'd prefer not to comment on all aspects of an upcoming release until it is finalized, but suffice to say that much of the changes are meant to streamline usage, fix bugs, and to increase compatibility with Neper.

**Reply by rquey** [reply]
FYI, I'm adding the displacement output to Neper -S right now. Will comment later.

**Reply by mkasemer** [reply]
Likewise, I am adding `disp` as output in FEPX as well, to be available in next release. Displacement output will be total (cumulative) displacements as x, y, z components. Is this acceptable @CsatiZoltan?

**Reply by CsatiZoltan** [reply]
@mkasemer Yes, that will be perfect.

**Reply by rquey** [reply]
Okay, `-noderes disp` is now live in Neper (which will still be useful if it was not among FEPX's output). (not documented yet)

@CsatiZoltan It should be possible to get a functional Neper without admin access, although it is certainly no fun. You would have to install the dependencies locally and link them. Most Neper dependencies are optional and can be switched off. Only GSL is mandatory but it can be compiled easily (it has no dependency). The next release of Neper will have some of its dependencies built-in to ease installation.

**Reply by CsatiZoltan** [reply]
Thank you, @rquey 
As I wanted to perform the visualization on the cluster (no need to move results from the cluster to the local machine, and also Neper is parallelized so the visualization is faster), I needed the GSL and the POV-Ray dependencies. As `conda` is available on the university cluster (as part of the Anaconda distribution), I could make use of it. This is how I installed Neper:

1. Activate the Anaconda module on the cluster
   ```bash
   $ module load anaconda3/2020.02
   ```
2. Create a conda environment from the following file (*environment.yml*)
   ```
   name: Neper
   channels:
   - conda-forge
   dependencies:
   - povray
   - neper
   ```
   ```bash
   $ conda env create -f environment.yml
   ```
3. Activate that environment and check if Neper is indeed installed locally
   ```bash
   $ conda activate Neper
   $ which neper
   ~/.conda/envs/Neper/bin/neper
   ```

I hope it helps others. The downsides of this approach:
- it doesn't use the latest version of Neper: https://anaconda.org/conda-forge/neper
- the created environment takes more than 2 GB of space (not a big issue on a cluster, though)

**Reply by mkasemer** [reply]
@CsatiZoltan: please note that the release was again delayed (more bugs!) but that it is coming imminently. The Neper fix should work for the interim.

**Reply by mkasemer** [reply]
@CsatiZoltan: [FEPX v1.2.0](https://github.com/acmelab-ua/FEPX/releases/tag/v1.2.0) has been released, which contain some updates to output (among other things). Please note that you may need to update to [Neper v4.2.0](https://github.com/rquey/neper/releases/tag/v4.2.0) for full compatibility.

**Reply by CsatiZoltan** [reply]
@mkasemer Wow, great news. Big thanks to the FEPX and Neper developers!
